great_expectations/experimental/datasources/dynamic_pandas.py

```diff
@@ -0,0 +1,284 @@
+    """
+    # TODO: parse the annotation string
+    annotation = param.annotation
+    if not isinstance(annotation, str):
+        logger.debug(type(annotation), annotation)
+        return annotation
+
+    types: list = []
+
+    union_parts = annotation.split(" | ")
+    str_to_eval: str
+    # TODO: use eval'ed types and use subclass check
+    for type_str in union_parts:
+        if type_str in CAN_HANDLE:
+            types.append(type_str)
+        else:
+            NEED_SPECIAL_HANDLING[param.name].add(type_str)
+    if not types:
+        return UNSUPPORTED_TYPE
+    if len(types) > 1:
+        str_to_eval = f"Union[{', '.join(types)}]"
+    else:
+        str_to_eval = types[0]
+    return str_to_eval
+
+
+def _to_pydantic_fields(
+    sig_tuple: _SignatureTuple,
+) -> Dict[str, _FieldSpec]:
+    """
+    Extract the parameter details in a structure that can be easily unpacked to
+    `pydantic.create_model()` as field arguments
+    """
+    fields_dict: Dict[str, _FieldSpec] = {}
+    for param_name, param in sig_tuple.signature.parameters.items():
+
+        no_annotation: bool = param.annotation is inspect._empty
+        if no_annotation:
+            logger.debug(f"`{param_name}` has no type annotation")
+            FIELD_SKIPPED_NO_ANNOTATION.add(param_name)  # TODO: not skipped
+            type_ = Any
+        else:
+            type_ = _get_annotation_type(param)
+            if type_ is UNSUPPORTED_TYPE or type_ == "None":
+                logger.debug(f"`{param_name}` has no supported types. Field skipped")
+                FIELD_SKIPPED_UNSUPPORTED_TYPE.add(param_name)
+                continue
+
+        substitution = FIELD_SUBSTITUTIONS.get(param_name)
+        if substitution:
+            fields_dict.update(substitution)
+        else:
+            fields_dict[param_name] = _FieldSpec(
+                type=type_, default_value=_get_default_value(param)  # type: ignore[arg-type]
+            )
+
+    return fields_dict
+
+
+M = TypeVar("M", bound=DataAsset)
+
+
+def _create_pandas_asset_model(
+    model_name: str,
+    model_base: Type[M],
+    type_field: Tuple[Union[Type, str], str],
+    fields_dict: Dict[str, _FieldSpec],
+) -> Type[M]:
+    """https://docs.pydantic.dev/usage/models/#dynamic-model-creation"""
+    model = pydantic.create_model(model_name, __base__=model_base, type=type_field, **fields_dict)  # type: ignore[call-overload] # FieldSpec is a tuple
+    return model
+
+
+def _generate_data_asset_models(
+    base_model_class: Type[M], whitelist: Optional[Sequence[str]] = None
+) -> Dict[str, Type[M]]:
+    io_methods = _extract_io_methods(whitelist)
+    io_method_sigs = _extract_io_signatures(io_methods)
+
+    models: Dict[str, Type[M]] = {}
+    for signature_tuple in io_method_sigs:
+
+        fields = _to_pydantic_fields(signature_tuple)
+
+        type_name = signature_tuple.name.split("read_")[1]
+        model_name = _METHOD_TO_CLASS_NAME_MAPPINGS.get(
```

Comment:
We can, I was doing this to preserve the uppercase for abbreviations because I thought that was our convention.

https://github.com/great-expectations/great_expectations/pull/6780#discussion_r1089017142